libcontainer/configs/validate/validator.go

```diff
@@ -108,6 +110,27 @@ func (v *ConfigValidator) usernamespace(config *configs.Config) error {
 		if _, err := os.Stat("/proc/self/ns/user"); os.IsNotExist(err) {
 			return fmt.Errorf("USER namespaces aren't enabled in the kernel")
 		}
+
+		root, err := os.Stat("/")
+		if err != nil {
+			return fmt.Errorf("Cannot stat root to check if the process is inside a chroot")
+		}
+
+		// Verify that the current process is not inside a chroot,
+		// the kernel does not allow creating a new user namespace in a chroot.
+		// for information, see: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/kernel/user_namespace.c
+		if root.Sys().(*syscall.Stat_t).Ino != 2 {
+			return fmt.Errorf("USER namespaces can't be created inside a chroot")
+		}
+
+		// On debian kernels there is specific patch to add a sysctl entry to enable user namespaces since
+		// the kernel is compiled without user namespaces by default.
+		// See:  kernel.unprivileged_userns_clone
+		unc, err := ioutil.ReadFile("/proc/sys/kernel/unprivileged_userns_clone")
+		if err == nil && unc[0] == '0' && os.Getuid() != 0 {
```

Comment:
I'm not a huge fan of distro-specific code (especially since Red Hat have a similar patch but it has different semantics and they also have some other patches as well). However, this is also not correct if the system has `/usr/bin/newuidmap` and `/usr/bin/newgidmap` set up -- see #1529.

Reply:
Agree